src/components/course-listing.cjsx

```diff
@@ -0,0 +1,76 @@
+_ = require 'underscore'
+React = require 'react'
+BS = require 'react-bootstrap'
+Router = require 'react-router'
+
+{CourseListingActions, CourseListingStore} = require '../flux/course-listing'
+
+# Called once the store is loaded
+# checks the course and roles and will redirect if there is only a single course and role
+DisplayOrRedirect = (transition, callback) ->
+  courses = CourseListingStore.allCourses() or []
+  if courses.length is 1 and courses[0].roles?.length is 1
+    roleType = courses[0].roles[0].type
+    type = switch roleType
+      when 'student' then 'viewStudentDashboard'
+      when 'teacher' then 'taskplans'
+      else
+        throw new Error("BUG: Unrecognized role type #{roleType}")
+    transition.redirect(type, {courseId: _.first(courses).id})
+  callback()
+
+
+CourseListing = React.createClass
+  displayName: 'CourseListing'
+
+  contextTypes:
+    router: React.PropTypes.func
+
+  statics:
+    # Called before the Router mounts and renders the component
+    # Uses the callback to delay rendering until the CourseListingStore is loaded
+    # and then calls DisplayOrRedirect above to perhaps redirect to a different component
+    willTransitionTo: (transition, params, query, callback) ->
+      unless CourseListingStore.isLoaded()
+        CourseListingActions.load() unless CourseListingStore.isLoading()
+        CourseListingStore.once 'loaded', ->
```

Comment:
You could do `CourseListingStore.once 'failed', -> callback` and in `render: ->` display the error message `unless CourseListingStore.isLoaded()`. Agreed there's no loading message though